pkg/operator/quorumguardcleanup/quorumguardcleanupcontroller.go

```diff
@@ -0,0 +1,234 @@
+}
+
+func (c *QuorumGuardCleanupController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.cleanupOldQuorumGuard(ctx, syncCtx)
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "QuorumGuardCleanupControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "Error",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("QuorumGuardCleanupControllerUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "QuorumGuardCleanupControllerDegraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+// cleanupOldQuorumGuard checks to see if the new static controller guard pods are ready
+// and then deletes the "etcd-quorum-guard" deployment and PodDisruptionBudget
+func (c *QuorumGuardCleanupController) cleanupOldQuorumGuard(ctx context.Context, syncCtx factory.SyncContext) error {
+	isTargetVersion, err := c.checkClusterOperatorVersion()
+	if err != nil {
+		return fmt.Errorf("failed to ensure clusteroperator version: %w", err)
+	}
+	if !isTargetVersion {
+		syncCtx.Queue().AddAfter(syncCtx.QueueKey(), 1*time.Minute)
+		return nil
+	}
+
+	newGuardLabelSelector, err := labels.Parse(newGuardLabelSelectorString)
+	if err != nil {
+		return fmt.Errorf("failed to parse label selector: %w", err)
+	}
+	newGuardPods, err := c.podLister.Pods(operatorclient.TargetNamespace).List(newGuardLabelSelector)
+	if err != nil {
+		return fmt.Errorf("failed to get etcd static controller guard pods: %w", err)
+	}
+	masterNodeLabelSelector, err := labels.Parse(masterNodeLabelSelectorString)
+	if err != nil {
+		return fmt.Errorf("failed to parse label selector: %w", err)
+	}
+	masterNodes, err := c.masterNodeLister.List(masterNodeLabelSelector)
+	if err != nil {
+		return fmt.Errorf("failed to list master nodes: %w", err)
+	}
+
+	// Wait until all new guard pods are rolled out to all master nodes, and have the ready condition
+	numReadyPods := countReadyPods(newGuardPods)
+	if numReadyPods != len(masterNodes) {
+		klog.V(2).Infof("%d/%d guard pods ready. Waiting until all new guard pods are ready", numReadyPods, len(newGuardPods))
+		syncCtx.Queue().AddAfter(syncCtx.QueueKey(), 1*time.Minute)
+		return nil
+	}
+
+	if err := c.removeDeployment(ctx, syncCtx); err != nil {
+		return err
+	}
+	if err := c.removePDB(ctx, syncCtx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// checkClusterOperatorVersion ensures the operator version in clusteroperator/etcd matches the desired version
+// in clusterversion/version
+// i.e all operands have been rolled out to the new image
+func (c *QuorumGuardCleanupController) checkClusterOperatorVersion() (bool, error) {
```

Comment:
>You need a check to be sure that the current level of this clusteroperator's version matches the desired level. Otherwise when this operator starts during an upgrade, it will immediately remove quorum guard protection.

> oh, I see it down below by looking for the actual pods. I think that's more complicated, but if you prefer it, ok.

@deads2k Sorry, something is off with github comments and I didn't see your [second comment](https://github.com/openshift/cluster-etcd-operator/pull/763/files#r834334470) until just now.

I've already added this check for the clusteroperator version but just double checking if this is what you meant by checking the desired level. I'm a bit unfamiliar with how the clusteroperator is updating the "operator" version but I think it waits for the operand versions to roll out.
https://github.com/openshift/cluster-etcd-operator/blob/f3d8db9c071c979b064063e97206f24398e9a854/vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/staticpodstate/staticpodstate_controller.go#L143-L158

And yes I already have the check for making sure the new guard pods are all ready before the old ones are cleaned up as well.